### ISSUE TITLE
How to get rid of all abstract classes, generated in the Models section?

### ISSUE THREAD

**Issue by MagicMaxxx** [RicoSuter/NSwag#2998]
Hello!

ASP.NET Core 3.1, WebAPI. 

I have just about 8 methods (like Login, Register... Ping.. and so on), but services.AddSwaggerDocument(..) generates about 50+ classes in the swagger document (https://localhost:44398/swagger/index.html), below in the Models section. These classes came from the whole solution and also there are a lot of abstract classes over here. 

Question is - how to control, what class should be there in that section? I planned to have just method parameters there...

Thanks!

**Comment by jeremyVignelles** [comment]
You need to be a bit more precise. Your localhost help won't help (we can't access your localhost...).

If many classes gets included, that usually means that you set a parameter or a return type to a type that was not meant to be serialized.

NSwag isn't supposed to generate 50 classes for simple cases like yours. If those advices didn't help, please share a minimal code that reproduces your issue.

**Comment by MagicMaxxx** [comment]
> You need to be a bit more precise. Your localhost help won't help (we can't access your localhost...).
> 
> If many classes gets included, that usually means that you set a parameter or a return type to a type that was not meant to be serialized.
> 
> NSwag isn't supposed to generate 50 classes for simple cases like yours. If those advices didn't help, please share a minimal code that reproduces your issue.

Well, I've updated my question above, and right after that, I've got your response - thanks for the answer! 

I understand that you can't access my localhost link - it just a sample of URL where I found Models section :)

Here is a piece of code I tried to use in the Startup.cs:

            services.AddSwaggerDocument(opt =>
            {
                DefaultContractResolver resolver = opt.ActualContractResolver as DefaultContractResolver;
                //resolver.SerializeCompilerGeneratedMembers = false;
                //JsonSerializerSettings settings = opt.ActualSerializerSettings;

                //opt.RequireParametersWithoutDefault = true;
                opt.GenerateAbstractSchemas = false;
                opt.GenerateAbstractProperties = false;
                //opt.GenerateExamples = false;
                //opt.SchemaGenerator.Settings.

                opt.AddOperationFilter(ctx =>
                {
                    return true;
                });
         }
but here I can't figure out how to filter all that Models that are in the Models list on the Swagger page... The goal of my question is to get rid of all abstract classes and classes that are not parameters/results of WebAPI methods...

**Comment by MagicMaxxx** [comment]
![Annotation 2020-08-12 184101](https://user-images.githubusercontent.com/19684558/90036202-6b9acb80-dccb-11ea-854c-6fdb3d37a167.png)

**Comment by jeremyVignelles** [comment]
Are these the base classes of your models ? In your APIs, you should use DTO objects (objects that are made specifically for the json serialization/deserialization), and not your internal objects, otherwise NSwag will try to serialize the complete object hierarchy.

**Comment by MagicMaxxx** [comment]
I don't think that this is a good idea to show (at least) abstract classes...  It would be good to have some method to control Models like you control Operations:
            services.AddSwaggerDocument(opt =>
            {
                opt.AddOperationFilter(ctx =>
                {
                    return false;
                });
            }
Here, if you return FALSE - there will be no Operations (methods in WebAPI) at all...

**Comment by jeremyVignelles** [comment]
You could edit the generated items, but I think there's value in generating every abstract class in the hierarchy.

Say you have the Apple and Orange classes, based on the Fruit abstract class.
How would you declare your `Eat(Fruit)` method without that abstract class? Would you duplicate the `Weight` property on both Apple and Orange ?

You have two solutions for your specific use case. Either you use a [DocumentProcessor](https://github.com/RicoSuter/NSwag/wiki/Document-Processors-and-Operation-Processors), or you clean up your API by only using serializable DTO objects without a base class, and copy the values in your internal classes manually.

**Comment by RicoSuter** [comment]
You can flatten all inheritance chains with FlattenInheritanceHierarchy config: 

https://apimundo.com/organizations/nuget-org/nuget-feeds/public/packages/NJsonSchema/versions/latest?tab=types&type=NJsonSchema.Generation.JsonSchemaGeneratorSettings

or you add the [JsonSchemaFlatten] attribute to the classes which should not have base classes: 

https://github.com/RicoSuter/NJsonSchema/blob/master/src/NJsonSchema/Annotations/JsonSchemaFlattenAttribute.cs
